File: ios/LivingRoomEdge/LivingRoomEdge/Brain/BrainEndpoint.swift
import Foundation
import Network

/// Two persisted Brain slots (LAN / Cloud) plus helpers to turn a base into `/api/v1/intent`.
/// Fresh-install defaults come from `config/endpoints.json` via `python3 tools/sync_endpoints.py`.
enum BrainEndpoint {
    /// LAN identity (settings). HTTP uses `lanResolvedBaseURL` (IPv4 from mDNS/probe).
    static let defaultLanBase = "http://brain.local:9527"
    static let defaultCloudBase = "http://115.190.153.53:9527"

    static let defaultLanIntentURL = intentURL(from: defaultLanBase)
    static let defaultCloudIntentURL = intentURL(from: defaultCloudBase)

    private static let lanKey = "livingroom.brain.lanURL"
    private static let cloudKey = "livingroom.brain.cloudURL"
    private static let lanResolvedKey = "livingroom.brain.lanResolvedBase"
    private static let identityMigratedKey = "livingroom.brain.lanIdentityMigrated"

    enum Mode: String {
        case lan
        case cloud

        var displayName: String { self == .lan ? "局域网" : "云端" }
    }

    /// How the phone chooses between the two saved Brain slots.
    enum Routing: String, CaseIterable, Identifiable {
        case auto
        case lan
        case cloud

        var id: String { rawValue }

        var pickerLabel: String { title }

        var title: String {
            switch self {
            case .auto: return "按网络自动"
            case .lan: return "锁定局域网"
            case .cloud: return "锁定云端"
            }
        }

        var subtitle: String {
            switch self {
            case .auto:
                return "在家 Wi‑Fi 且局域网 Brain 可达时走局域网，否则走云端。"
            case .lan:
                return "始终连家里的 Brain。外出或局域网不通时对话会失败。"
            case .cloud:
                return "始终连云端 Brain。即使在家也不走局域网。"
            }
        }
    }

    private static let routingKey = "livingroom.brain.routing"

    static var routing: Routing {
        get {
            let raw = UserDefaults.standard.string(forKey: routingKey) ?? ""
            return Routing(rawValue: raw) ?? .auto
        }
        set { UserDefaults.standard.set(newValue.rawValue, forKey: routingKey) }
    }

    static var lanBaseURL: String {
        get { stored(key: lanKey, fallback: defaultLanBase) }
        set { UserDefaults.standard.set(normalizeBase(newValue), forKey: lanKey) }
    }

    /// IPv4 Brain base used for HTTP (`http://192.168.x.x:9527`). Empty until discovery/probe.
    static var lanResolvedBaseURL: String {
        get { stored(key: lanResolvedKey, fallback: "") }
        set {
            let next = normalizeBase(newValue)
            if next.isEmpty {
                UserDefaults.standard.removeObject(forKey: lanResolvedKey)
            } else {
                UserDefaults.standard.set(next, forKey: lanResolvedKey)
            }
        }
    }

    static var cloudBaseURL: String {
        get { stored(key: cloudKey, fallback: defaultCloudBase) }
        set { UserDefaults.standard.set(normalizeBase(newValue), forKey: cloudKey) }
    }

    /// URL actually used to talk to LAN Brain. Empty until an IPv4 is discovered — never `brain.local`.
    static var lanConnectBaseURL: String {
        migrateLanIdentityIfNeeded()
        if let ip = ipv4Base(from: lanResolvedBaseURL) { return ip }
        if let ip = ipv4Base(from: lanBaseURL) { return ip }
        return ""
    }

    static var lanIntentURL: String { intentURL(from: lanConnectBaseURL) }
    static var cloudIntentURL: String { intentURL(from: cloudBaseURL) }

    static func migrateLanIdentityIfNeeded() {
        guard !UserDefaults.standard.bool(forKey: identityMigratedKey) else {
            if ipv4Host(from: lanBaseURL) != nil, lanResolvedBaseURL.isEmpty {
                lanResolvedBaseURL = lanBaseURL
                lanBaseURL = defaultLanBase
            }
            return
        }
        if let ipBase = ipv4Base(from: lanBaseURL) {
            if lanResolvedBaseURL.isEmpty {
                lanResolvedBaseURL = ipBase
            }
            lanBaseURL = defaultLanBase
        } else if lanBaseURL.isEmpty {
            lanBaseURL = defaultLanBase
        }
        UserDefaults.standard.set(true, forKey: identityMigratedKey)
    }

    static func ipv4Host(from raw: String) -> String? {
        let trimmed = normalizeBase(raw)
        let host: String
        if let url = URL(string: trimmed), let urlHost = url.host, !urlHost.isEmpty {
            host = urlHost
        } else {
            host = trimmed
        }
        let parts = host.split(separator: ".")
        guard parts.count == 4, parts.allSatisfy({ UInt8($0) != nil }) else { return nil }
        return host
    }

    /// `brain.local` / `gateway.local` — Bonjour names. Must not be used as HTTP/TCP host.
    static func isBonjourHost(_ raw: String) -> Bool {
        let trimmed = normalizeBase(raw)
        let host: String
        if let url = URL(string: trimmed), let urlHost = url.host, !urlHost.isEmpty {
            host = urlHost
        } else {
            host = trimmed
        }
        let lower = host.lowercased()
        return lower.hasSuffix(".local") || lower.contains(".local.")
    }

    /// Fail-fast copy for NSURLSession: never send `.local` over HTTP.
    static func refuseBonjourHTTP(_ raw: String) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return "没有可用地址" }
        if isBonjourHost(trimmed) {
            return "拒绝用 mDNS 名发 HTTP（\(trimmed)）。必须先发现 IPv4。"
        }
        return nil
    }

    /// Distinguish DNS/mDNS failures from TCP/HTTP timeout (both are NSURLErrorDomain).
    static func describeTransportError(_ error: Error, url: URL, elapsed: String) -> String {
        let ns = error as NSError
        let host = url.host ?? "?"
        let viaIP = ipv4Host(from: url.absoluteString) != nil
        let target = viaIP ? "IPv4 \(host)" : "主机名 \(host)"
        let where_ = url.absoluteString
        guard ns.domain == NSURLErrorDomain else {
            return "\(ns.localizedDescription) · \(target) · \(where_) · \(elapsed)"
        }
        switch ns.code {
        case NSURLErrorTimedOut:
            return "请求超时（不是 DNS）· \(target) · \(where_) · \(elapsed)"
        case NSURLErrorCannotFindHost, NSURLErrorDNSLookupFailed:
            return "DNS/mDNS 解析失败 · \(target) · \(where_) · \(elapsed)"
        case NSURLErrorCannotConnectToHost:
            return "TCP 连不上 · \(target) · \(where_) · \(elapsed)"
        case NSURLErrorNetworkConnectionLost:
            return "连接中断 · \(target) · \(where_) · \(elapsed)"
        case NSURLErrorNotConnectedToInternet:
            return "无网络 · \(target) · \(where_) · \(elapsed)"
        default:
            return "NSURLError \(ns.code) \(ns.localizedDescription) · \(target) · \(where_) · \(elapsed)"
        }
    }

    static func ipv4Base(from raw: String) -> String? {
        guard let host = ipv4Host(from: raw) else { return nil }
        let port: Int
        if let url = URL(string: normalizeBase(raw)), let urlPort = url.port {
            port = urlPort
        } else {
            port = 9527
        }
        return "http://\(host):\(port)"
    }

    static func intentURL(from raw: String) -> String {
        let base = normalizeBase(raw)
        guard !base.isEmpty else { return "" }
        if isBonjourHost(base) { return "" }
        if base.hasSuffix("/api/v1/intent") { return base }
        if base.contains("/api/v1/") { return base }
        return base + "/api/v1/intent"
    }

    static func displayBase(from raw: String) -> String {
        var value = normalizeBase(raw)
        if value.hasSuffix("/api/v1/intent") {
            value = String(value.dropLast("/api/v1/intent".count))
        }
        return normalizeBase(value)
    }

    static func normalizeBase(_ raw: String) -> String {
        var value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        while value.hasSuffix("/") {
            value.removeLast()
        }
        return value
    }

    private static func stored(key: String, fallback: String) -> String {
        let saved = UserDefaults.standard.string(forKey: key)?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return saved.isEmpty ? fallback : normalizeBase(saved)
    }
}

enum BrainPathKind: String {
    case wifi
    case wired
    case cellular
    case none
    case unknown

    var looksOnHomeLAN: Bool {
        self == .wifi || self == .wired
    }

    var label: String {
        switch self {
        case .wifi: return "Wi‑Fi"
        case .wired: return "有线"
        case .cellular: return "蜂窝"
        case .none: return "无网络"
        case .unknown: return "未知"
        }
    }
}

struct BrainNetworkEnvironment: Equatable {
    var pathKind: BrainPathKind = .unknown
    var looksOnHomeLAN: Bool = false
    var lanProbeOk: Bool?
    var lanProbeDetail: String = ""
    var routing: BrainEndpoint.Routing = .auto
    var mode: BrainEndpoint.Mode = .cloud
    var activeIntentURL: String = BrainEndpoint.defaultCloudIntentURL

    var modeLabel: String { mode.displayName }

    var routingLabel: String { routing.title }

    var activeBaseURL: String { BrainEndpoint.displayBase(from: activeIntentURL) }

    var summaryLine: String {
        let path = looksOnHomeLAN ? "像在家庭局域网（\(pathKind.label)）" : "不在家庭局域网（\(pathKind.label)）"
        let probe: String
        switch lanProbeOk {
        case true: probe = "LAN 探测成功"
        case false: probe = "LAN 探测失败"
        case nil: probe = "尚未探测 LAN"
        }
        return "\(path) · \(probe) · 路由 \(routingLabel) · 当前 \(modeLabel) \(activeBaseURL)"
    }
}

File: android/living-room-android/src/main/java/com/smarthome/livingroom_android/brain/BrainEndpoint.kt
package com.smarthome.livingroom_android.brain

/**
 * Two persisted Brain slots (LAN / Cloud) plus helpers to turn a base into `/api/v1/intent`.
 * Mirrors iOS `BrainEndpoint`.
 * Fresh-install defaults: edit `config/endpoints.json` then `python3 tools/sync_endpoints.py`.
 */
object BrainEndpoint {
    const val DEFAULT_LAN_BASE = "http://192.168.3.73:9527"
    const val DEFAULT_CLOUD_BASE = "http://115.190.153.53:9527"

    enum class Mode {
        LAN,
        CLOUD,
        ;

        val label: String get() = if (this == LAN) "局域网" else "云端"
    }

    enum class Routing {
        AUTO,
        LAN,
        CLOUD,
        ;

        val title: String
            get() = when (this) {
                AUTO -> "按网络自动"
                LAN -> "锁定局域网"
                CLOUD -> "锁定云端"
            }

        val pickerLabel: String get() = title

        val subtitle: String
            get() = when (this) {
                AUTO -> "在家 Wi‑Fi 且局域网 Brain 可达时走局域网，否则走云端。"
                LAN -> "始终连家里的 Brain。外出或局域网不通时对话会失败。"
                CLOUD -> "始终连云端 Brain。即使在家也不走局域网。"
            }

        val wire: String
            get() = when (this) {
                AUTO -> "auto"
                LAN -> "lan"
                CLOUD -> "cloud"
            }

        companion object {
            fun fromWire(raw: String?): Routing {
                return when (raw?.trim()?.lowercase()) {
                    "lan" -> LAN
                    "cloud" -> CLOUD
                    else -> AUTO
                }
            }
        }
    }

    enum class PathKind {
        WIFI,
        WIRED,
        CELLULAR,
        NONE,
        UNKNOWN,
        ;

        val looksOnHomeLAN: Boolean get() = this == WIFI || this == WIRED

        val label: String
            get() = when (this) {
                WIFI -> "Wi‑Fi"
                WIRED -> "有线"
                CELLULAR -> "蜂窝"
                NONE -> "无网络"
                UNKNOWN -> "未知"
            }
    }

    fun normalizeBase(raw: String): String {
        var value = raw.trim()
        while (value.endsWith("/")) {
            value = value.dropLast(1)
        }
        return value
    }

    fun intentUrl(from: String): String {
        val base = normalizeBase(from)
        if (base.endsWith("/api/v1/intent")) return base
        if (base.contains("/api/v1/")) return base
        return "$base/api/v1/intent"
    }

    fun displayBase(from: String): String {
        var value = normalizeBase(from)
        if (value.endsWith("/api/v1/intent")) {
            value = value.dropLast("/api/v1/intent".length)
        }
        return normalizeBase(value)
    }

    fun apiUrl(fromIntentOrBase: String, leaf: String): String {
        val trimmed = normalizeBase(fromIntentOrBase)
        val path = when {
            trimmed.endsWith("/intent") ->
                trimmed.dropLast("intent".length) + leaf.trimStart('/')
            trimmed.contains("/api/v1/") -> {
                val idx = trimmed.indexOf("/api/v1/")
                trimmed.substring(0, idx + "/api/v1/".length) + leaf.trimStart('/')
            }
            else -> "$trimmed/api/v1/${leaf.trimStart('/')}"
        }
        return path
    }

    fun destLabel(intentUrl: String, cloudBase: String = DEFAULT_CLOUD_BASE): String {
        val given = displayBase(intentUrl)
        val cloud = displayBase(cloudBase)
        if (given.equals(cloud, ignoreCase = true)) return "cloud"
        val givenHost = runCatching { java.net.URI(given).host }.getOrNull()?.lowercase().orEmpty()
        val cloudHost = runCatching { java.net.URI(cloud).host }.getOrNull()?.lowercase().orEmpty()
        if (givenHost.isNotEmpty() && givenHost == cloudHost) return "cloud"
        return "img_server"
    }

    fun pingUrl(fromIntentOrBase: String, clientTimeMs: Long): String {
        val base = apiUrl(fromIntentOrBase, "ping")
        val sep = if (base.contains("?")) "&" else "?"
        return "${base}${sep}client_time_ms=$clientTimeMs"
    }

    fun intentsListUrl(
        fromIntentOrBase: String,
        participantId: String,
        beforeId: Int?,
        limit: Int,
    ): String {
        val base = apiUrl(fromIntentOrBase, "intents")
        val builder = StringBuilder(base)
        builder.append("?participant_id=").append(java.net.URLEncoder.encode(participantId, "UTF-8"))
        builder.append("&limit=").append(limit.coerceIn(1, 5))
        if (beforeId != null && beforeId >= 1) {
            builder.append("&before_id=").append(beforeId)
        }
        return builder.toString()
    }

    fun intentDetailUrl(fromIntentOrBase: String, intentId: String): String {
        val base = apiUrl(fromIntentOrBase, "intent_detail")
        val sep = if (base.contains("?")) "&" else "?"
        return "${base}${sep}intent_id=${java.net.URLEncoder.encode(intentId, "UTF-8")}"
    }

    fun assetContentUrl(
        fromIntentOrBase: String,
        assetId: String,
        intentId: String,
        representation: String,
    ): String {
        val encodedId = java.net.URLEncoder.encode(assetId, "UTF-8")
        val base = apiUrl(fromIntentOrBase, "assets/$encodedId/content")
        val sep = if (base.contains("?")) "&" else "?"
        val iid = java.net.URLEncoder.encode(intentId, "UTF-8")
        val rep = java.net.URLEncoder.encode(representation, "UTF-8")
        return "${base}${sep}intent_id=$iid&representation=$rep"
    }

    fun livingRoomIntentsPullUrl(fromIntentOrBase: String): String =
        apiUrl(fromIntentOrBase, "devices/living-room/intents")
}

data class BrainNetworkEnvironment(
    val pathKind: BrainEndpoint.PathKind = BrainEndpoint.PathKind.UNKNOWN,
    val looksOnHomeLAN: Boolean = false,
    val lanProbeOk: Boolean? = null,
    val lanProbeDetail: String = "",
    val routing: BrainEndpoint.Routing = BrainEndpoint.Routing.AUTO,
    val mode: BrainEndpoint.Mode = BrainEndpoint.Mode.CLOUD,
    val activeIntentUrl: String = BrainEndpoint.intentUrl(BrainEndpoint.DEFAULT_CLOUD_BASE),
    val resolveBusy: Boolean = false,
) {
    val modeLabel: String get() = mode.label
    val routingLabel: String get() = routing.title
    val activeBaseUrl: String get() = BrainEndpoint.displayBase(activeIntentUrl)

    val summaryLine: String
        get() {
            val path = if (looksOnHomeLAN) {
                "像在家庭局域网（${pathKind.label}）"
            } else {
                "不在家庭局域网（${pathKind.label}）"
            }
            val probe = when (lanProbeOk) {
                true -> "LAN 探测成功"
                false -> "LAN 探测失败"
                null -> "尚未探测 LAN"
            }
            return "$path · $probe · 路由 $routingLabel · 当前 $modeLabel $activeBaseUrl"
        }
}
